src/Evolution/Systems/GrMhd/ValenciaDivClean/ConservativeFromPrimitive.cpp

```diff
@@ -39,17 +39,47 @@ void ConservativeFromPrimitive::apply(
     const Scalar<DataVector>& sqrt_det_spatial_metric,
     const tnsr::ii<DataVector, 3, Frame::Inertial>& spatial_metric,
     const Scalar<DataVector>& divergence_cleaning_field) noexcept {
-  const auto spatial_velocity_one_form =
-      raise_or_lower_index(spatial_velocity, spatial_metric);
-  const auto magnetic_field_one_form =
-      raise_or_lower_index(magnetic_field, spatial_metric);
-  const auto magnetic_field_dot_spatial_velocity =
-      dot_product(magnetic_field, spatial_velocity_one_form);
-  const auto spatial_velocity_squared =
-      dot_product(spatial_velocity, spatial_velocity_one_form);
-  // not const to save an allocation later
-  auto magnetic_field_squared =
-      dot_product(magnetic_field, magnetic_field_one_form);
+  Variables<tmpl::list<
+      hydro::Tags::SpatialVelocityOneForm<DataVector, 3, Frame::Inertial>,
+      hydro::Tags::SpatialVelocitySquared<DataVector>,
+      hydro::Tags::MagneticFieldOneForm<DataVector, 3, Frame::Inertial>,
+      hydro::Tags::MagneticFieldDotSpatialVelocity<DataVector>,
+      hydro::Tags::MagneticFieldSquared<DataVector>>>
+      temp_tensors{get(rest_mass_density).size()};
+  const auto& spatial_velocity_one_form =
+      get<hydro::Tags::SpatialVelocityOneForm<DataVector, 3, Frame::Inertial>>(
+          temp_tensors);
+  raise_or_lower_index(
```

Comment:
why not mutating references that can be used in the `make_not_null` calls?

Reply:
Sorry, I was previously doing this so we would have more const things but people seem to prefer mutating references from code review, so I will update.

Reply:
Done and rebased on latest develop